crates/miden-tx/src/host/mod.rs

```diff
@@ -509,23 +527,97 @@ where
     ///
-    /// Expected stack state: `[event, NOTE_METADATA, RECIPIENT, ...]`
+    /// Expected stack state: `[event, NOTE_METADATA, note_ptr, RECIPIENT, note_idx]`
+    ///
+    /// This method extracts recipient data from the advice provider and attempts to build a
+    /// NoteRecipient. If the note script is not present in the advice provider, it returns
+    /// [`TransactionEventHandling::Unhandled`] to request the script from the data store.
     fn on_note_after_created(
         &mut self,
         process: &ProcessState,
-    ) -> Result<(), TransactionKernelError> {
-        let stack = process.get_stack_state().split_off(1);
-        // # => [NOTE_METADATA]
+    ) -> Result<TransactionEventHandling, TransactionKernelError> {
+        // Extract metadata from stack
+        let metadata_word = process.get_stack_word(1);
+        let metadata = NoteMetadata::try_from(metadata_word)
+            .map_err(TransactionKernelError::MalformedNoteMetadata)?;
 
-        let note_idx: usize = stack[9].as_int() as usize;
+        // Extract recipient digest from stack
+        let recipient_digest = process.get_stack_word(6);
 
-        assert_eq!(note_idx, self.output_notes.len(), "note index mismatch");
+        // Extract note index from stack
+        let note_idx = process.get_stack_item(10).as_int() as usize;
 
-        let note_builder = OutputNoteBuilder::new(stack, process.advice_provider())?;
+        // Check if the advice provider contains the recipient data
+        let recipient = if let Some(data) =
+            process.advice_provider().get_mapped_values(&recipient_digest)
+        {
+            // NoteDetails from AdviceProvider: [num_inputs, INPUTS_COMMITMENT, SCRIPT_ROOT,
+            // SERIAL_NUM]
+            if data.len() != 13 {
+                return Err(TransactionKernelError::MalformedRecipientData(data.to_vec()));
+            }
 
-        self.output_notes.insert(note_idx, note_builder);
+            let num_inputs = data[0].as_int() as usize;
+            let inputs_commitment = Word::new([data[1], data[2], data[3], data[4]]);
+            let script_root = Word::new([data[5], data[6], data[7], data[8]]);
+            let serial_num = Word::from([data[9], data[10], data[11], data[12]]);
+
+            // Check if the script is in the advice provider
+            let script_data = process.advice_provider().get_mapped_values(&script_root);
+
+            let Some(script_data) = script_data else {
+                // Script is missing - request it from the data store
+                return Ok(TransactionEventHandling::Unhandled(TransactionEventData::NoteScript {
+                    script_root,
+                    metadata,
+                    recipient_digest,
+                    note_idx,
+                }));
+            };
 
-        Ok(())
+            // Script is present, build the recipient
+            let inputs_data = process.advice_provider().get_mapped_values(&inputs_commitment);
```

Comment:
Related to one of the above comments: this is the other occurrence of the code that I think could be encapsulated into a helper function.